### PR TITLE
Update flutter/http.dart to use dart:io.

### PR DESCRIPTION
**Pull request by chinmaygarde**
- Updates Flutter's copy of http.dart to match tip-of-tree with the use of mirrors removed.
- Removes unused NetworkAssetBundle and MojoAssetBundle.fromNetwork.
- We can now deprecate the Mojo network stack.

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Fixes https://github.com/flutter/flutter/issues/5808
Fixes https://github.com/flutter/flutter/issues/5668
Replaces https://github.com/flutter/engine/pull/3008

cc @abarth @jason-simmons

**Comment by chinmaygarde** [comment]
Linters and the media service example needs to be updated. Patching.

File: packages/flutter/lib/http.dart [hunk 1]
```diff
@@ -3,7 +3,7 @@
 // found in the LICENSE file.
 
 /// A [Future]-based library for making HTTP requests.
-/// 
+///
 /// To use, import `package:flutter/http.dart`.
 ///
 /// This library is based on Dart's `http` package, but differs in that it is a
```
Comment (Hixie) [inline packages/flutter/lib/http.dart]: comment needs updating :-)

File: packages/flutter/lib/src/http/base_client.dart [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+/// The abstract base class for an HTTP client. This is a mixin-style class;
+/// subclasses only need to implement [send] and maybe [close], and then they
+/// get various convenience methods for free.
+abstract class BaseClient implements Client {
+  /// Sends an HTTP HEAD request with the given headers to the given URL, which
+  /// can be a [Uri] or a [String].
+  ///
+  /// For more fine-grained control over the request, use [send] instead.
+  Future<Response> head(url, {Map<String, String> headers}) =>
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: nit: style guide says not to break lines when using =>

Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: also, spaces inside { }

File: packages/flutter/lib/src/http/base_client.dart [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  /// state of the stream; it could have data written to it asynchronously at a
+  /// later point, or it could already be closed when it's returned. Any
+  /// internal HTTP errors should be wrapped as [ClientException]s.
+  Future<StreamedResponse> send(BaseRequest request);
+
+  /// Sends a non-streaming [Request] and returns a non-streaming [Response].
+  Future<Response> _sendUnstreamed(String method, url,
+      Map<String, String> headers, [body, Encoding encoding]) async {
+
+    if (url is String) url = Uri.parse(url);
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: nit: if body on its own line

File: packages/flutter/lib/src/http/base_client.dart [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    }
+
+    return Response.fromStream(await send(request));
+  }
+
+  /// Throws an error if [response] is not successful.
+  void _checkResponseSuccess(url, Response response) {
+    if (response.statusCode < 400) return;
+    var message = "Request to $url failed with status ${response.statusCode}";
+    if (response.reasonPhrase != null) {
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: nit: no braces around single-line statements

File: packages/flutter/lib/src/http/base_client.dart [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      }
+    }
+
+    return Response.fromStream(await send(request));
+  }
+
+  /// Throws an error if [response] is not successful.
+  void _checkResponseSuccess(url, Response response) {
+    if (response.statusCode < 400) return;
+    var message = "Request to $url failed with status ${response.statusCode}";
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: nit: single quotes for strings

File: packages/flutter/lib/src/http/base_client.dart [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+      message = "$message: ${response.reasonPhrase}";
+    }
+    if (url is String) url = Uri.parse(url);
+    throw new ClientException("$message.", url);
+  }
+
+  /// Closes the client and cleans up any resources associated with it. It's
+  /// important to close each client when it's done being used; failing to do so
+  /// can cause the Dart process to hang.
+  void close() {}
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_client.dart]: nit: empty code blocks should have a space inside them

File: packages/flutter/lib/src/http/base_request.dart [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+  /// Whether the request has been finalized.
+  bool get finalized => _finalized;
+  bool _finalized = false;
+
+  /// Creates a new HTTP request.
+  BaseRequest(this.method, this.url)
+    : headers = new LinkedHashMap(
+        equals: (key1, key2) => key1.toLowerCase() == key2.toLowerCase(),
+        hashCode: (key) => key.toLowerCase().hashCode);
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_request.dart]: nit: closing ) should be on its own line if opening ( is on its own line

File: packages/flutter/lib/src/http/base_request.dart [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      return new StreamedResponse(
+          new ByteStream(stream),
+          response.statusCode,
+          contentLength: response.contentLength,
+          request: response.request,
+          headers: response.headers,
+          isRedirect: response.isRedirect,
+          persistentConnection: response.persistentConnection,
+          reasonPhrase: response.reasonPhrase);
+    } catch (_) {
```
Comment (Hixie) [inline packages/flutter/lib/src/http/base_request.dart]: nit: no _

File: packages/flutter/lib/src/http/mojo_client.dart [hunk 9]
```diff
@@ -1,260 +0,0 @@
...
-    loader.start(request, (mojom.UrlResponse response) async {
-      loader.close();
-      try {
-        if (response.error != null)
-          throw new Exception('Request to "$url" failed with error ${response.error.code}.\n${response.error.description}');
-        if (!response.body.handle.isValid)
-          throw new Exception('Response body does not have a valid handle, but no error was reported.\n${response.body}');
-        completer.complete(response);
-      } catch (e, stack) {
-        FlutterError.reportError(new FlutterErrorDetails(
```
Comment (Hixie) [inline packages/flutter/lib/src/http/mojo_client.dart]: losing this error handling makes me sad

**Comment by Hixie** [comment]
(I only mentioned each nit once, but many of them apply to many files.)

Is there no way we can just reuse the code instead of forking it? This is a whole lot of code that none of us know and will have trouble maintaining.

**Comment by chinmaygarde** [comment]
> Is there no way we can just reuse the code instead of forking it? This is a whole lot of code that none of us know and will have trouble maintaining.

There were a ton of updates upstream and we had gotten rid of everything except the MojoClient when we first imported it in. So it is hard to figure out exactly where to patch in the new dart:io client. Besides, this has been tested upstream. We have only made minor changes to get rid of mirrors.

@abarth did mention that this entire package was a stopgap measure till we deprecate it completely. The number of linter errors I would have to manually fix makes me really montivated to just deprecate the thing now and use `dart:io` internally. Can we do this? That way, we really wont have anything to maintain.

**Comment by sethladd** [comment]
> Can we do this?

From a product perspective, removing it from the Flutter SDK is fine. I'd like to see us collaborate with the Dart team on how someone could publish a package that is intended to work in/on Flutter.

**Comment by chinmaygarde** [comment]
https://github.com/dart-lang/http/issues/1#issuecomment-246788519 seems to indicate that the fix is near. But I am not sure what the blockers are, or, if there is an ETA. I am going to move this to packages/ and import it by path. When that issue is fixed, we can import it from pub. Not going to manually apply the Flutter style guide to that codebase.

**Comment by eseidelGoogle** [comment]
I spoke with @dgrove.  His guidance was that the fix for dart-lang/http#1 is around 3 weeks off.  I would like us to land this CL (or something like it) as an interim step to unblock customer: fast.

**Comment by chinmaygarde** [comment]
We can host the patched http package on Github and add a Git dependency in the pubspec. Once the actual package is published in three weeks, we update the dependency to point to package `http` actual. Can I have flutter/http on GH for this? I can also point to my mirror https://github.com/chinmaygarde/flutter_http if thats OK.

**Comment by eseidelGoogle** [comment]
@chinmaygarde and I discussed that a separate package would add unnecessary complexity to our integration with Google's internal build systems.  Any change we make here is intended to be a temporary (few weeks) fix until http/http.dart proper is fixed.

**Comment by chinmaygarde** [comment]
In that case, may I land this without updating the external package to match the Flutter style guide? Seems impractical to go through the effort and have it be removed in a couple of weeks.

**Comment by abarth** [comment]
Yes, I don't think its worth changing the style.  If this issue drags on for more than a month, we can reconsider the style issue.

**Comment by chinmaygarde** [comment]
@Hixie mentioned in chat that we could publish a new package called `temporary_http_for_transition_period_do_not_use` with a descriptive README and use that for a few weeks.

Importing this code into the tree seems to be problematic for the bots as I am not able to find a way to make the analyzer ignore certain paths. It currently fails with about 200 issues (mostly missing annotations, using `as` etc.).

**Comment by abarth** [comment]
That seems fine to me.

**Comment by eseidelGoogle** [comment]
Can we tell the analyzer to ignore this directory?
https://github.com/flutter/flutter/blob/master/.analysis_options#L29
If we go the separate-package route we'll need to update our internal systems more (which is doable, just we'll need to coordinate with that team).

**Comment by chinmaygarde** [comment]
I tried that but it didn't seems to do anything. Maybe I am holding it wrong though. I just noticed there are multiple `.analysis_options` files. Could have edited the wrong one.

**Comment by eseidelGoogle** [comment]
@danrubel or @Hixie could tell you which do which, and presumably if it's non-obvious we should update the comments in the top of each to explain which to use.

**Comment by chinmaygarde** [comment]
I think I may be hitting https://github.com/dart-lang/sdk/issues/26212 where the excludes are ignored.

**Comment by chinmaygarde** [comment]
@tvolkert: How does publishing our own temporary package affect the internal build system (if at all)?

**Comment by chinmaygarde** [comment]
I am not able to re-open this issue. https://github.com/flutter/flutter/pull/5851 is the replacement with the linter fixes.

**Pull request closed without merging**